addons/website/static/src/snippets/s_website_form/000.js

```diff
@@ -126,18 +126,26 @@ odoo.define('website.s_website_form', function (require) {
             // Because, using t-att- inside form make it non-editable
             // Data-fill-with attribute is given during registry and is used by
             // to know which user data should be used to prfill fields.
+            const inputsWithoutValueProperty = ['button', 'hidden', 'image', 'reset', 'submit',
+                                                'checkbox', 'radio'];
             var $dataFor = $('[data-for=' + this.$target.attr('id') + ']');
             if ($dataFor.length || !_.isEmpty(this.preFillValues)) {
-                var dataForValues = $dataFor.length ? JSON.parse($dataFor.data('values').replace('False', '""').replace('None', '""').replace(/'/g, '"')) : [];
+                this.dataForValues = $dataFor.length ? JSON.parse($dataFor.data('values').replace('False', '""').replace('None', '""').replace(/'/g, '"')) : [];
                 var fields = _.pluck(this.$target.serializeArray(), 'name');
                 _.each(fields, function (field) {
                     var $field = self.$target.find('input[name="' + field + '"], textarea[name="' + field + '"]');
-                    if (!$field.val() && _.has(dataForValues, field) && dataForValues[field]) {
-                        $field.val(dataForValues[field]);
-                    } else if (!$field.val() && $field.data('fill-with')) {
+                    const inputType = $field.prop('type');
+                    if ($field.prop('tagName') === 'INPUT' &&
+                    inputsWithoutValueProperty.includes(inputType)) {
+                        // In order to keep the DOM intact, we can't fill value
+                        // attribute, only value property.
+                        return;
+                    }
```

Comment:
- Is there a real need to check all those types? Is not the problematic one only "hidden"? Do we currently allow prefilling on checkboxes or worse: submit buttons?
- The comment is not bad but is not explicit enough to me:
    - "keep the DOM intact", replace "intact" by "untouched"
    - "we can't fill value attribute", well you can, you *should* not... but why? It needs explanations.
    - Someone reading this might say "omg it's bad, we cannot prefill hidden fields"... you should explain that it is done on send
    => Something like "Avoid altering the DOM for the pre-filling of input values (interact only with input value *property*) so that those are not accidentally saved during edition. The pre-filling of hidden inputs (which forces a DOM alteration) will only be done on form submit." ? 

Actually, reading that version, I wonder if we can do that as a stable fix: a form with a hidden input as no more value during page life if it was prefilled before... custo may rely on this. It may actually even be bad for master. When I suggested to do this, I had in mind hidden inputs had a value property too, I did not expect the solution would lead to have hidden input with undefined values up until they are send. Your original solution (with many more comments) is probably better then (just setting `website_form_original_default_value` when prefilled via data-for or prefill rpc)). Actually the code which removes those values https://github.com/odoo/odoo/commit/b637a5e32f767b62736241042f88fa0cecf9f10b#diff-ec70bde754e307ff6e9c3f50d61e703606fc40619e066464c99ffaddfdbb90aeR439 -> I am still not sure why it is in `cleanForSave` rather than in the publicWidget destroy though.

So: if you agree, restore what you did before but explain all this with inline code comments :+1: Sorry for the back and forth, that's the R in R&D I guess :grimacing: